### PR TITLE
feat: Use regex parsing for dsn

### PR DESCRIPTION
**Pull request by stepansergeevitch**
Updated DSN parser to use regexp under the hood to add support for special symbols in password

### CODE REVIEW

File: dsn.go [hunk 1]
```diff
@@ -13,73 +14,43 @@ type fireboltSettings struct {
...
-func parseRemainingDSN(dsn string, expectedPrefix string, stopChars []uint8) (string, string, error) {
-	if !strings.HasPrefix(dsn, expectedPrefix) {
-		return "", dsn, fmt.Errorf("expected prefix not found: %s", expectedPrefix)
-	}
-
-	dsn = dsn[len(expectedPrefix):]
-	res, dsn := splitString(dsn, stopChars)
-	return res, dsn, nil
-}
+const dsnPattern = `^firebolt://(?P<username>.*@.*):(?P<password>.*)@(?P<database>\w+)(?:/(?P<engine>[^?]+))?(?:\?(?P<parameters>\w+\=[^=&]+(?:\&\w+=[^=&]+)*))?$`
```
Comment (ptiurin) [inline dsn.go:17]: Nice one. Tried it out and it seems to work well. One thing - we need to make sure parameters don't contain `@` or parsing gets messed up.

File: dsn.go [hunk 2]
```diff
@@ -13,73 +14,43 @@ type fireboltSettings struct {
...
-	result.database, dsn, err = parseRemainingDSN(dsn, "@", []uint8{'/', '?'})
-	if err != nil {
-		return nil, ConstructNestedError("error during database parsing", err)
+	result.username = dsnMatch[1]
+	result.password = dsnMatch[2]
+	result.database = dsnMatch[3]
+	if len(dsnMatch[4]) > 0 {
+		// engine name was provided
+		result.engineName = dsnMatch[4]
 	}
```
Comment (ptiurin) [inline dsn.go:41]: Is there a way to leverage named capture groups in the regex for access here? e.g. `dsnMatch.username` would be safer than `dsnMatch[1]`

File: dsn.go
Comment (ptiurin) [inline dsn.go:1]: How does the service account auth work here?

File: dsn.go (hunk 2 above)
Comment (stepansergeevitch) [reply dsn.go:41]: matcher provider names as a separate list. We would still need to access the value by index

File: dsn.go
Comment (stepansergeevitch) [reply dsn.go:1]: Good point, this one was failing. Fixed now

**Review by ptiurin — APPROVED** [review]

**Pull request merged**